src/blueprint/from_config.rs

```diff
@@ -207,7 +208,101 @@ fn to_fields(type_of: &config::Type, config: &Config) -> Valid<Vec<blueprint::Fi
+where
+  F: Fn(&'a str) -> Option<&FieldDefinition>,
+{
+  match head {
+    "value" => {
+      if let Some(val) = get_value(tail) {
+        if val.of_type.is_nullable() {
+          return Valid::fail(format!("value '{tail}' is a nullable type"));
+        }
+      } else {
+        return Valid::fail(format!("no value '{tail}' found"));
+      }
+    }
+    // TODO this should probably error out on the `ListType` match arm since it
+    // doesn't make sense to have a list in the template URL
+    "args" => {
+      // XXX this is a linear search but it's cost is less than that of
+      // constructing a HashMap since we'd have 3-4 arguments at max in
+      // most cases
+      if let Some(arg) = args.iter().find(|arg| arg.name == tail) {
+        if arg.default_value.is_none() && arg.of_type.is_nullable() {
+          return Valid::fail(format!("argument '{tail}' is a nullable type"));
+        }
+      } else {
+        return Valid::fail(format!("no argument '{tail}' found"));
+      }
+    }
+    "header" | "vars" => {
+      // TODO neither of these values make sense for usage in a URL template
+      // should we error out here?
+
+      // "header" refers to the header values known at runtime, which we can't
+      // validate here
+
+      // "vars" refer to the server's configuration variables, set up in
+      // `config/server.rs` and they all have fallbacks to default values, so
+      // we don't need any checks here
+    }
+    _ => {
+      // TODO have we covered all the cases? Does reaching here mean some
+      // invariant was violated?
+    }
+  }
+
+  Valid::Ok(())
+}
+
+fn validate_fields(fields: &[FieldDefinition]) -> Valid<()> {
+  let mut validation_map = HashMap::<&str, &FieldDefinition>::new();
+  validation_map.reserve(fields.len());
+
+  for field in fields {
+    if validation_map.insert(&field.name, field).is_some() {
+      // TODO is this really unreachable? Do we already error out on duplicated
+      // fields?
+      panic!("Field '{}' shouldn't be already present!", field.name);
+    }
+  }
+
+  for field in fields {
+    // XXX we could use `Mustache`'s `render` method with a mock
+    // struct implementing the `PathString` trait encapsulating `validation_map`
+    // but `render` simply falls back to the default value for a given
+    // type if it doesn't exist, so we wouldn't be able to get enough
+    // context from that method alone
+    // So we must duplicate some of that logic here :(
+    if let Some(Expression::Unsafe(Operation::Endpoint(req_template, _, _))) = &field.resolver {
+      match &req_template.root_url {
+        Mustache(segments) => {
+          for segment in segments {
+            if let Segment::Expression(parts) = segment {
+              // TODO is this invariant enforced anywhere so that we can omit
+              // this check?
+              if parts.len() < 2 {
+                continue;
```

Comment:
```
  user: User! @http(path: "/users/{{value.userId}}")
  test: User @http(path: "/users/{{value.user.phone}}")
```

@tusharmath This is valid, right? Looks like will have to refactor a bit here to account for nested fields